### PR TITLE
[ tool ] Fix expression evaluation not handling errors correctly

### LINKED ISSUE #157922: Expression evaluation returns true (instead of an error) for "a is X" when X is an unresolved identifier
### Steps to reproduce

Moving from https://github.com/dart-lang/sdk/issues/56300 based on @bkonyi's comments.

---

@FMorschel reported to me conditional breakpoints not working as expected, and we tracked it back to expression evaluation returning `true` for conditions that we didn't expect. We managed to reproduce it with just simple expression eval:

image

The error messages shown here only show up because I'm running in `-verbose` mode, otherwise you'd only see the `true` responses. When doing `x is A` where `A` is some class that's apparently not in scope, the result is just `true`. For non-Flutter dart, you get a reasonable error:

image

The VM traffic is below. The `flutter run` `stdout` output shows the error in verbose mode, but the VM Service response to `evaluateInFrame` returns a valid boolean of `true` with no indication that the request failed.

### Expected results

There should be an error message, like in Dart, so it's clear the condition failed and didn't evaluate to `true`.

### Actual results

The expression returns as if it evaluated to `true`.

### Code sample

<details open><summary>Code sample</summary>

```dart
// Evaluate this in the Debug Console while running a Flutter app
a is SomeClassThatDoesNotExist
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

image

image

</details>

### Logs

<details open><summary>Logs</summary>

```console
==> [VM] {
	"jsonrpc": "2.0",
	"id": "741",
	"method": "evaluateInFrame",
	"params": {
		"isolateId": "isolates/3676968675006663",
		"frameIndex": 0,
		"expression": "a is FutureOr<int>",
		"disableBreakpoints": true
	}
}
<== [Flutter] [+44567 ms] org-dartlang-debug:synthetic_debug_expression:1:6: Error: 'FutureOr' isn't a type.
<== [Flutter] [        ] a is FutureOr<int>
<== [Flutter] [        ]      ^^^^^^^^
<== [VM] {
	"jsonrpc": "2.0",
	"result": {
		"type": "@Instance",
		"_vmType": "Bool",
		"class": {
			"type": "@Class",
			"fixedId": true,
			"id": "classes/62",
			"name": "bool",
			"location": {
				"type": "SourceLocation",
				"script": {
					"type": "@Script",
					"fixedId": true,
					"id": "libraries/@0150898/scripts/dart%3Acore%2Fbool.dart/0",
					"uri": "dart:core/bool.dart",
					"_kind": "kernel"
				},
				"tokenPos": 508,
				"endTokenPos": 9302,
				"line": 15,
				"column": 1
			},
			"library": {
				"type": "@Library",
				"fixedId": true,
				"id": "libraries/@0150898",
				"name": "dart.core",
				"uri": "dart:core"
			}
		},
		"identityHashCode": 0,
		"kind": "Bool",
		"fixedId": true,
		"id": "objects/bool-true",
		"valueAsString": "true"
	},
	"id": "741"
}
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[√] Flutter (Channel master, 3.27.0-1.0.pre.51, on Microsoft Windows [Version 10.0.22631.4317], locale en-GB)
    • Flutter version 3.27.0-1.0.pre.51 on channel master at D:\Dev\Google\Flutter\Flutter main
    • Upstream repository git@github.com:DanTup/flutter.git
    • FLUTTER_GIT_URL = git@github.com:DanTup/flutter.git
    • Framework revision 6829c6a9b9 (2 weeks ago), 2024-10-15 22:36:36 +0800
    • Engine revision 107ea8baa4
    • Dart version 3.6.0 (build 3.6.0-334.2.beta)
    • DevTools version 2.40.1

[√] Windows Version (Installed version of Windows is version 10 or higher)

[√] Android toolchain - develop for Android devices (Android SDK version 35.0.0-rc3)
    • Android SDK at C:\Users\danny\AppData\Local\Android\sdk
    • Platform android-35, build-tools 35.0.0-rc3
    • Java binary at: C:\Program Files\Android\Android Studio\jbr\bin\java
    • Java version OpenJDK Runtime Environment (build 17.0.9+0--11185874)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop Windows apps (Visual Studio Community 2022 17.10.5)
    • Visual Studio at C:\Program Files\Microsoft Visual Studio\2022\Community
    • Visual Studio Community 2022 version 17.10.35122.118
    • Windows 10 SDK version 10.0.22621.0

[√] Android Studio (version 2023.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.9+0--11185874)

[√] VS Code (version 1.95.0)
    • VS Code at C:\Users\danny\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.101.20241031

[√] VS Code (version 1.93.0-insider)
    • VS Code at C:\Users\danny\AppData\Local\Programs\Microsoft VS Code Insiders
    • Flutter extension version 3.95.20240801

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.22631.4317]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 130.0.6723.91
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 127.0.2651.86

[√] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by bkonyi**
Also starts DDS for integration test devices.

Fixes https://github.com/flutter/flutter/issues/157922

### CODE REVIEW

**Comment by bkonyi** [comment]
FYI @DanTup

**Comment by andrewkolos** [comment]
(FYI that I probably won't have time to look at this for at least a couple of days)

**Comment by bkonyi** [comment]
> (FYI that I probably won't have time to look at this for at least a couple of days. Let me know if anything else is blocked on this PR)

No rush, nothing's blocked on this :)

File: packages/flutter_tools/lib/src/run_hot.dart [hunk 1]
```diff
@@ -212,8 +212,10 @@ class HotRunner extends ResidentRunner {
             await device.generator!.compileExpression(expression, definitions,
                 definitionTypes, typeDefinitions, typeBounds, typeDefaults,
                 libraryUri, klass, method, isStatic);
-        if (compilerOutput != null && compilerOutput.expressionData != null) {
+        if (compilerOutput != null && compilerOutput.errorCount == 0 && compilerOutput.expressionData != null) {
           return base64.encode(compilerOutput.expressionData!);
+        } else if (compilerOutput!.errorCount > 0 && compilerOutput.errorMessage != null) {
```
Comment (navaronbracke) [inline packages/flutter_tools/lib/src/run_hot.dart:217]: Don't you need a null check here, like on line 215?

```
if (compilerOutput != null && compilerOutput.errorCount == 0 && compilerOutput.expressionData != null) {
  return base64.encode(compilerOutput.expressionData!);
} else if (compilerOutput != null && compilerOutput.errorCount > 0 && compilerOutput.errorMessage != null) {
  throw VmServiceExpressionCompilationException(compilerOutput.errorMessage!);
}
```

File: packages/flutter_tools/lib/src/test/flutter_platform.dart [hunk 2]
```diff
@@ -422,8 +423,10 @@ class FlutterPlatform extends PlatformPlugin {
       await compiler!.compiler!.compileExpression(expression, definitions,
         definitionTypes, typeDefinitions, typeBounds, typeDefaults, libraryUri,
         klass, method, isStatic);
-    if (compilerOutput != null && compilerOutput.expressionData != null) {
+    if (compilerOutput != null && compilerOutput.errorCount == 0 && compilerOutput.expressionData != null) {
```
Comment (navaronbracke) [inline packages/flutter_tools/lib/src/test/flutter_platform.dart:426]: Same remark here

File: packages/flutter_tools/lib/src/run_hot.dart (hunk 1 above)
Comment (bkonyi) [reply packages/flutter_tools/lib/src/run_hot.dart]: Good catch! Done.

File: packages/flutter_tools/lib/src/test/flutter_platform.dart (hunk 2 above)
Comment (bkonyi) [reply packages/flutter_tools/lib/src/test/flutter_platform.dart]: Done.

File: packages/flutter_tools/test/integration.shard/expression_evaluation_test.dart [hunk 3]
```diff
@@ -43,52 +43,46 @@ void batch1() {
...
     await breakInBuildMethod(flutter);
     await evaluateComplexReturningExpressions(flutter);
-    await cleanProject();
+  });
+
+    testWithoutContext('evaluating invalid expressions throws an exception with compilation error details', () async {
+    await flutter.run(withDebugger: true);
+    await breakInTopLevelFunction(flutter);
+    await evaluateInvalidExpression(flutter);
   });
```
Comment (andrewkolos) [inline packages/flutter_tools/test/integration.shard/expression_evaluation_test.dart:86]:
```suggestion
      await flutter.run(withDebugger: true);
      await breakInTopLevelFunction(flutter);
      await evaluateInvalidExpression(flutter);
```

**Review by andrewkolos — APPROVED** [review]
LGTM with nit

File: packages/flutter_tools/test/integration.shard/expression_evaluation_test.dart (hunk 3 above)
Comment (andrewkolos) [reply packages/flutter_tools/test/integration.shard/expression_evaluation_test.dart:86]: I'm still figuring out how to do suggestions from VSCode properly, so this is wrong. I mean to try to fix the indentation here

Comment (andrewkolos) [reply packages/flutter_tools/test/integration.shard/expression_evaluation_test.dart:86]: second try:
```suggestion
    await flutter.run(withDebugger: true);
    await breakInTopLevelFunction(flutter);
    await evaluateInvalidExpression(flutter);
```

Comment (bkonyi) [reply packages/flutter_tools/test/integration.shard/expression_evaluation_test.dart:86]: I'll just update it myself... :)

**Pull request merged**